### PR TITLE
0.3.0

### PR DESCRIPTION
**Pull request by DrVilepis**
Yep

### CODE REVIEW

File: src/api/oauth.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    request: Query<TokenExchangeRequest>,
+    client: Data<Client>,
+    db: Data<DbPool>,
+) -> Result<impl Responder, TimeError> {
+    if request.code.chars().any(|c| !c.is_alphanumeric()) {
+        return Err(TimeError::BadCode);
+    }
+
+    let res = client
+        .post("http://localhost:6969/api/v1/token")
```
Comment (lajp) [inline src/api/oauth.rs:61]: Suspicious hard-coded url

File: src/api/oauth.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            ("client_secret", &CLIENT_INFO.secret),
+        ]))
+        .await
+        .unwrap()
+        .json::<TokenResponse>()
+        .await
+        .unwrap();
+
+    let res = client
+        .get("http://localhost:6969/api/v1/me")
```
Comment (lajp) [inline src/api/oauth.rs:76]: Suspicious hard-coded url

File: src/api/oauth.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        .unwrap()
+        .json::<TestausIdApiUser>()
+        .await
+        .unwrap();
+
+    let token =
+        block(move || testausid_login(&mut db.get()?, res.id, res.name, res.platform.id)).await??;
+
+    Ok(HttpResponse::PermanentRedirect()
+        .insert_header(("location", "https://testaustime.fi/oauth_redirect"))
```
Comment (lajp) [inline src/api/oauth.rs:89]: Suspicious hard-coded url

File: src/api/oauth.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        .cookie(
+            Cookie::build("testaustime_token", token)
+                .domain("testaustime.fi")
```
Comment (lajp) [inline src/api/oauth.rs:92]: Suspicious hard-coded domain

File: src/api/search.rs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+use crate::{database, error::TimeError, DbPool};
+
+#[derive(Deserialize)]
+pub struct UserSearch {
+    pub keyword: String,
+}
+
+//TODO: Maybe return small coding summary?
+//FIXME: The error when missing `keyword` is ugly, go fix
```
Comment (lajp) [inline src/api/search.rs:15]: this is fixed afaik

File: src/api/users.rs [hunk 6]
```diff
@@ -59,34 +62,83 @@ pub async fn get_activities(
...
-        let friend_id = get_user_by_name(&conn, &path.0)?.id;
-        if friend_id == user.id {
-            let conn = db.get()?;
-            let data = block(move || get_activity(&conn, data.into_inner(), friend_id)).await??;
-            Ok(web::Json(data))
+        //FIXME: This is technically not required when the username equals the username of the
+        //authenticated user
+        let target_user = database::get_user_by_name(&mut conn, &path.0)?;
+
+        if target_user.id == user.id || target_user.is_public {
```
Comment (lajp) [inline src/api/users.rs:74]: Add the else – if as another || to get rid of duplicate code

File: src/api/users.rs [hunk 7]
```diff
@@ -59,34 +62,83 @@ pub async fn get_activities(
...
+    user: UserId,
+    db: Data<DbPool>,
+) -> Result<impl Responder, TimeError> {
+    let mut conn = db.get()?;
+    let data = if path.0 == "@me" {
+        block(move || database::get_all_activity(&mut conn, user.id)).await??
+    } else {
+        let target_user = database::get_user_by_name(&mut conn, &path.0)?;
+
+        if target_user.id == user.id || target_user.is_public {
```
Comment (lajp) [inline src/api/users.rs:102]: Add the else – if as another || to get rid of duplicate code

File: src/main.rs [hunk 8]
```diff
@@ -1,7 +1,7 @@
 #![feature(let_else, once_cell)]
 
 mod api;
-mod database;
+pub mod database;
```
Comment (lajp) [inline src/main.rs:4]: remove `pub` from this and the other two lines underneath

**Review by lajp — CHANGES_REQUESTED** [review]
Some minor changes but all in all this is looking good.

**Review by lajp — DISMISSED** [review]
This is good. Nice work!

**Review by lajp — APPROVED** [review]
Go merg

**Comment by DrVilepis** [comment]
Will do

**Pull request merged**